libyul/Utilities.cpp

```diff
@@ -86,56 +86,96 @@ std::string solidity::yul::reindent(std::string const& _code)
+{
+	if (_literal == "true")
+		return LiteralValue{1, "true"};
+	else if (_literal == "false")
+		return LiteralValue{0, "false"};
 
-	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
-	if (isNew)
+	yulAssert(false, "Unexpected bool literal value!");
+}
+LiteralValue solidity::yul::valueOfLiteral(std::string_view const _literal, LiteralKind const& _kind, bool const _unlimitedLiteralArgument)
+{
+	switch (_kind)
 	{
-		std::string const& literalString = _literal.value.str();
-		yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-		it->second = u256(literalString);
+	case LiteralKind::Number:
+		return valueOfNumberLiteral(_literal);
+	case LiteralKind::Boolean:
+		return valueOfBoolLiteral(_literal);
+	case LiteralKind::String:
+		return _unlimitedLiteralArgument ? valueOfBuiltinStringLiteralArgument(_literal) : valueOfStringLiteral(_literal);
 	}
-	return it->second;
+	yulAssert(false);
 }
 
-u256 solidity::yul::valueOfStringLiteral(Literal const& _literal)
+std::string solidity::yul::formatLiteral(solidity::yul::Literal const& _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::String, "Expected string literal!");
-	yulAssert(_literal.value.str().size() <= 32, "Literal string too long!");
+	std::string result;
+	if (_literal.value.unlimited())
+		result = _literal.value.builtinStringLiteralValue();
+	else if (_literal.value.hint() && _literal.value.value() == valueOfLiteral(*_literal.value.hint(), _literal.kind).value())
+		result = *_literal.value.hint();
+	else
+		if (_literal.kind == LiteralKind::Boolean)
+		{
+			yulAssert(_literal.value.value() == 0 || _literal.value.value() == 1, "Could not format boolean literal");
+			result = _literal.value.value() == 1 ? "true" : "false";
+		}
+		else
+			result = _literal.value.value().str();
 
-	return u256(h256(_literal.value.str(), h256::FromBinary, h256::AlignLeft));
+	return result;
 }
 
-u256 yul::valueOfBoolLiteral(Literal const& _literal)
+bool solidity::yul::validStringLiteral(solidity::yul::Literal const& _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::Boolean, "Expected bool literal!");
-
-	if (_literal.value == "true"_yulstring)
-		return u256(1);
-	else if (_literal.value == "false"_yulstring)
-		return u256(0);
+	if (_literal.kind == LiteralKind::String)
+	{
+		auto const repr = [&]() {
+			if (_literal.value.hint())
+				return *_literal.value.hint();
+			return _literal.value.value().str();
+		};
+		return _literal.value.unlimited() || repr().size() <= 32;
+	}
+	return false;
+}
 
-	yulAssert(false, "Unexpected bool literal value!");
+bool solidity::yul::validNumberLiteral(solidity::yul::Literal const& _literal)
+{
+	if (_literal.kind == LiteralKind::Number && !_literal.value.unlimited())
+	{
+		if (_literal.value.hint())
+		{
+			auto const repr = *_literal.value.hint();
+			return (isValidDecimal(repr) || isValidHex(repr)) && bigint(repr) <= u256(-1);
+		}
+		return true;
+	}
+	return false;
```

Comment:
Just a general style suggestion for these functions: I think the code is easier to process if you deal with special cases first rather than leave them hanging until you reach the `else`. It makes the flow more linear. For me it makes reading it easier:
```suggestion
	if (!_literal.kind == LiteralKind::Number || _literal.value.unlimited())
		return false;

	if (!_literal.value.hint())
		return true;

	auto const repr = *_literal.value.hint();
	return (isValidDecimal(repr) || isValidHex(repr)) && bigint(repr) <= u256(-1);
```

Reply:
Similarly, in `formatLiteral()`, you could make it much less nested if the conditions returned immediately instead of storing the value in `result` only to return it at the end.

Reply:
yeah, fair, your way of writing it is indeed easier to digest:)